lib_eio/pool.ml

```diff
@@ -114,25 +113,44 @@ let run_with t f slot =
         f x
     end
   with
-  | r -> 
+  | r ->
     add t slot;
     r
   | exception ex ->
     let bt = Printexc.get_raw_backtrace () in
     add t slot;
     Printexc.raise_with_backtrace ex bt
 
-let use t f =
+(* Creates a fresh resource [x], runs [f x], then disposes of [x] *)
+let run_new_and_dispose t f =
+  let x = t.alloc () in
+  match f x with
+  | r ->
+    t.dispose x;
+    r
+  | exception ex ->
+    let bt = Printexc.get_raw_backtrace () in
+    t.dispose x;
+    Printexc.raise_with_backtrace ex bt
+
+let use t ?(never_block=false) f =
   let segment, cell = Q.next_suspend t.q in
   match Atomic.get cell with
   | Finished | Request _ -> assert false
   | Resource slot ->
     Atomic.set cell Finished;   (* Allow value to be GC'd *)
     run_with t f slot
   | In_transition ->
-      (* It would have been better if more resources were available.
-         If we still have capacity, add a new slot now. *)
-      maybe_add_slot t;
+    let current = Atomic.get t.slots in
+    match current < t.max_slots with
+    | false when never_block ->
+      (* We are at capacity, but cannot block.
+         Create a new resource to run f but don't add it to the pool. *)
+      Atomic.set cell Finished;
```

Comment:
This isn't quite right. The cell might contain a `Resource` by now. If so, we should use that.

(probably needs an `Atomic.exchange cell Finished` here)

Reply:
Thanks, I'll fix it this weekend if you don't get to it first

Reply:
Pushed. I have a remaining question for you that I'm keeping for the devs meeting, so I think merging this PR should wait until after the meeting.